### PR TITLE
BUG: creating dataframe from masked numpy array turns missing string …

### LINKED ISSUE #54706: BUG: creating dataframe from masked numpy array turns missing string values into NaN
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd
import numpy as np
import numpy.ma as ma

x = np.array(['a', 'b'])
mx = ma.masked_array(x, mask=[0, 1])
print(pd.DataFrame(mx))
```

### Issue Description

creating dataframe from masked numpy array turns missing string values into NaN

### Expected Behavior

For examples like this, where the masked array is of string dtype, I expect pandas to fill missing data with `None` instead of `NaN`.

### Installed Versions

<details>
<summary>Details</summary>

```
INSTALLED VERSIONS
------------------
commit           : 0f437949513225922d851e9581723d82120684a6
python           : 3.8.16.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 22.5.0
Version          : Darwin Kernel Version 22.5.0: Mon Apr 24 20:51:50 PDT 2023; root:xnu-8796.121.2~5/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 2.0.3
numpy            : 1.24.4
pytz             : 2023.3
dateutil         : 2.8.2
setuptools       : 67.8.0
pip              : 23.1.2
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : 8.12.2
pandas_datareader: None
bs4              : None
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : None
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : None
snappy           : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
zstandard        : None
tzdata           : 2023.3
qtpy             : None
pyqt5            : None
```

</details>

### PR DESCRIPTION
**Pull request by kshitiz305**
…values into NaN

- [x] closes #54706 (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by kshitiz305** [comment]
take

**Comment by mroeschke** [comment]
Could you address the failing tests?

**Comment by kshitiz305** [comment]
@jbrockmendel Hi I am facing some issue while testing the code with the following

 ` # GH#18463
            warnings.warn(
                f"Mismatched null-like values {a} and {b} found. In a future "
                "version, pandas equality-testing functions "
                "(e.g. assert_frame_equal) will consider these not-matching "
                "and raise.",
                FutureWarning,
                stacklevel=find_stack_level()`
                
                
causing the test to fail while running 
                
 `pytest pandas/tests/series/test_constructors.py`
 
 
 it has been 10 days but there has been no response on slack regarding the query. Kindly help

File: pandas/_libs/testing.pyx [hunk 1]
```diff
@@ -2,6 +2,7 @@ import cmath
 import math
 import warnings
 
+warnings.simplefilter(action="ignore", category=FutureWarning)
```
Comment (jbrockmendel) [inline pandas/_libs/testing.pyx:5]: definitely not

**Comment by jbrockmendel** [comment]
> [...] causing the test to fail while running
>
> pytest pandas/tests/series/test_constructors.py

That generally means you are doing something like `tm.assert_series_equal(result, expected)` and result/expected contain different NA-like values, e.g. one has `np.nan` while the other has `None`.  To avoid the warning, they need to match.

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

**Pull request closed without merging**